### DISCUSSION TITLE
Engine Optimizations

### DISCUSSION THREAD

**Discussion by assemblaj** [ikemen-engine/Ikemen-GO#1707] [Engine development]
Just making a thread to document small engine optimizations I find. 

Testing IWTBT Ryu 4v4 on a low end PC and taking profile snapshots when IWTBTRyu does the Hadoken Wave, which is arguably his most problematic move,  and the Projectile.clsn function would case a lot of trouble 
![iwaswrong](https://github.com/ikemen-engine/Ikemen-GO/assets/7599535/b61cee92-5787-4a11-8905-1d78b42cfc2d)
![iwaswrong2](https://github.com/ikemen-engine/Ikemen-GO/assets/7599535/3a0be239-5ebb-4d96-a6da-0d9b0ab456c2)

This turned out to be because of the range statement over all of the character's projectiles in that function.  I'm not exactly sure why.  Anyway, this was fixed by replacing the range with a regular for loop that access the projectile through a pointer: 

```go
	for i := 0; i < len(sys.chars) && p.hits >= 0; i++ {
		if len(sys.chars[i]) == 0 || i == playerNo {
			continue
		}
		prSlice := sys.projs[i]
		for j := 0; j < len(sys.projs[i]); j++ {
			pr := &prSlice[j]
                        // logic 
                }
         }
```
![afterfix](https://github.com/ikemen-engine/Ikemen-GO/assets/7599535/56727701-5eb6-44b2-900a-1153cddaa60b)

The Systen,charUpdate and System.action calls shown in the above image have issues with the exact same thing. I fixed it there too, but I haven't profiled it yet. 

Not exactly sure why the graphics side of things went ahead and ate up any kind of performance gains we would've seen from this lol, but I know for sure that that Hadoken Wave that IWTBT Ryu does is more likely to play out without frameskipping with this fix. 

In the future, it might be a worthwhile idea to create helper functions that iterate over each item in our asset slices and perform this behavior while taking a function to perform on each individual item, like for example this maybe: 

```go
func IterProjectile(char int, process func(pr *Projectile)) {
	prSlice := sys.projs[char]
	for j := 0; j < len(sys.projs[char]); j++ {
		pr := &prSlice[j]
		process(pr)
	}
}
```
To prevent having to copy/paste this logic everywhere.  Of course some loops require more complex logic such that this wouldn't be applicable, but I think something like this would work in most cases.

**Comment by potsmugen** [comment]
I was just looking at this and it looks like, while we already iterate over characters through pointers (charList), we do a lot of copying the Projectile and Explod data when we iterate over them.

Update: This is the January 22nd Nightly with 1000 projectiles:

![image](https://github.com/user-attachments/assets/62100369-5a5c-4e24-8cf6-19ce4dc2d254)

And this is after optimizing the usage of "range" in projectile loops:

![image](https://github.com/user-attachments/assets/44847727-613f-4272-8ebf-8d7b0c4496c6)

6 vs 35fps. Of course a scenario with 1000 projectiles won't happen very often, but this still seems worth doing. Couldn't extract the same gains from explods though, so those don't seem as bad.
